src/list.alp

```diff
@@ -0,0 +1,307 @@
+
+
+{-| Reverse a list.
+
+    reverse [1, 2, 3, 4] == [4, 3, 2, 1]
+-}
+val length 'a : fn (list 'a) -> int
+let length xs =
+  beam :erlang :length [xs] with
+  | result -> result
+
+test "length of empty list" =
+  assert.equal (length []) 0
+
+test "length of list" =
+  assert.equal (length [1, 2, 3]) 3
+
+
+{-| Concatenate a bunch of lists into a single list.
+
+    concat [] == []
+    concat [[1, 2], [3], [4, 5]] == [1, 2, 3, 4, 5]
+-}
+val concat 'a : fn (list (list 'a)) -> (list 'a)
+let concat xxs =
+  beam :lists :concat [xxs] with
+  | result -> result
+
+test "concat of empty list" =
+  assert.equal (concat []) []
+
+test "concat of lists" =
+  assert.equal
+    (concat [[1, 2], [3], [4, 5]])
+    [1, 2, 3, 4, 5]
+
+
+{-| Keep only elements that satisfy the predicate.
+
+      filter (fn x -> x > 4) [10, 2, 3, 4, 5, 6] == [10, 5, 6]
+-}
+val filter 'a : fn (fn 'a -> bool) (list 'a) -> list 'a
+let filter predicate xs =
+  beam :lists :filter [predicate, xs] with
+  | result -> result
+
+test "filter of empty list" =
+  let isBig x = x > 4 in
+  assert.equal (filter isBig []) []
+
+test "filter of list list" =
+  let isBig x = x > 4 in
+  assert.equal
+    (filter isBig [10, 2, 3, 4, 5, 6])
+    [10, 5, 6]
+
+
+{-| Put two lists together.
+
+      append [1, 1, 2] [3, 5, 8] == [1, 1, 2, 3, 5, 8]
+      append ["a", "b"] ["c"] == ["a", "b", "c"]
+
+-}
+val append 'a : fn (list 'a) (list 'a) -> (list 'a)
+let append xs ys =
+  beam :lists :append [xs, ys] with
+  | result -> result
+
+test "append [] [...]" =
+  assert.equal
+    (append [] [1, 2, 3])
+    [1, 2, 3]
+
+test "append [...] []" =
+  assert.equal
+    (append [1, 2, 3] [])
+    [1, 2, 3]
+
+test "append [...] [...]" =
+  assert.equal
+    (append [1, 2, 3] [4, 5, 6])
+    [1, 2, 3, 4, 5, 6]
+
+
+{-| Reduce a list from the right.
+
+      foldRight add 0 [1, 2, 3] == 6
+      foldRight append [1] [[2], [3], [4]] == [2, 3, 4, 1]
+
+    So `foldRight step initialValue [1, 2, 3]` is like saying:
+
+      initialValue
+        |> step 3
+        |> step 2
+        |> step 1
+-}
+-- val foldRight 'a 'b : fn (fn 'a 'b -> 'b) 'b (list 'a) -> 'b -- TODO: The compiler hangs when this is uncommented
+let foldRight func acc ls =
+  beam :lists :foldr [func, acc, ls] with
+  | result -> result
+
+test "foldRight of empty list" =
+  assert.equal
+    (foldRight append [] [])
+    []
+
+test "foldRight of list" =
+  assert.equal
+    (foldRight append [1] [[2], [3], [4]])
+    [2, 3, 4, 1]
+
+
+{-| Reduce a list from the left.
+    This function is tail recursive and generally preferred to `foldRight/3`
+
+      foldLeft add 0 [1, 2, 3] == 6
+      foldLeft append [1] [[2], [3], [4]] == [4, 3, 2, 1]
+
+    So `foldLeft step initialValue [1, 2, 3]` is like saying:
+
+      initialValue
+        |> step 1
+        |> step 2
+        |> step 3
+-}
+-- val foldLeft 'a 'b : fn (fn 'a 'b -> 'b) 'b (list 'a) -> 'b -- TODO: The compiler hangs when this is uncommented
```

Comment:
Sounds like another situation with recursive reference cells in the typer :(

Could you please file an Alpaca issue with a minimal module example?  Thanks!